File: offer/36/36.js
/**
 * // Definition for a Node.
 * function Node(val,left,right) {
 *    this.val = val;
 *    this.left = left;
 *    this.right = right;
 * };
 **/
let pre, head
let treeToDoublyList = function (root) {
    if (!root) {
        return null;
    }
    dfs(root);
    head.left = pre;
    pre.right = head;
    return head;
};

function dfs(treeNode) {
    if (!treeNode) return;
    dfs(treeNode.left);
    if (pre){
        pre.right = treeNode;
    }
    else{
        head = treeNode;
    }
    treeNode.left = pre;
    pre =treeNode;
    dfs(treeNode.right);
}

File: offer/36/36.go
package main

import (
	. "../../TreeNode"
)

func treeToDoublyList(root *TreeNode) *TreeNode {
	var pre, head *TreeNode
	var dfs func(node *TreeNode)
	dfs = func(node *TreeNode) {
		if node == nil {
			return
		}
		dfs(node.Left)
		if pre != nil {
			pre.Right = node
		} else {
			head = node
		}
		node.Left = pre
		pre = node
		dfs(node.Right)
	}

	dfs(root)
	head.Left = pre
	pre.Right = head
	return head
}

func main() {

}
